File: fiberplane-charts/src/Mondrian/generator/utils/attachSuggestionsToXAxis.ts
import type { Axis } from "../../types";
import type { Buckets } from "../types";
import { getTimeFromTimestamp } from "./getTimeFromTimestamp";

/**
 * Adds suggestions to the axis based on the position of the first bucket and
 * the interval between buckets.
 *
 * @note This function mutates its input axis.
 */
export function attachSuggestionsToXAxis(
  xAxis: Axis,
  buckets: Buckets<unknown>,
  interval: number,
) {
  if (interval <= 0) {
    return;
  }

  const firstBucketTime = getFirstBucketTime(buckets);
  if (!firstBucketTime) {
    return;
  }

  const suggestions = [];

  let suggestion = firstBucketTime;
  while (suggestion < xAxis.maxValue) {
    if (suggestion >= xAxis.minValue) {
      suggestions.push(suggestion);
    }

    suggestion += interval;
  }

  xAxis.tickSuggestions = suggestions;
}

function getFirstBucketTime(buckets: Buckets<unknown>): number | undefined {
  let firstBucketTimestamp: string | undefined;
  for (const timestamp of buckets.keys()) {
    if (!firstBucketTimestamp || timestamp < firstBucketTimestamp) {
      firstBucketTimestamp = timestamp;
    }
  }

  return firstBucketTimestamp
    ? getTimeFromTimestamp(firstBucketTimestamp)
    : undefined;
}

File: mondrian-charts/src/utils/attach_suggestions_to_x_axis.rs
use super::get_time_from_timestamp;
use crate::types::{Axis, Buckets};

/// Adds suggestions to the axis based on the position of the first bucket and
/// the interval between buckets.
pub(crate) fn attach_suggestions_to_x_axis<T>(
    x_axis: &mut Axis,
    buckets: &Buckets<T>,
    interval: f64,
) {
    if interval <= 0. {
        return;
    }

    let Some(mut suggestion) = get_first_bucket_time(buckets) else {
        return;
    };

    let mut suggestions = Vec::new();

    while suggestion < x_axis.max_value {
        if suggestion >= x_axis.min_value {
            suggestions.push(suggestion);
        }

        suggestion += interval;
    }

    x_axis.tick_suggestions = Some(suggestions);
}

fn get_first_bucket_time<T>(buckets: &Buckets<T>) -> Option<f64> {
    let mut first_bucket_timestamp = None;
    for timestamp in buckets.keys() {
        if first_bucket_timestamp
            .map(|first_bucket_timestamp| timestamp < first_bucket_timestamp)
            .unwrap_or(true)
        {
            first_bucket_timestamp = Some(timestamp);
        }
    }

    first_bucket_timestamp.cloned().map(get_time_from_timestamp)
}
